From 3092ab7b9b743f41963522a6c36f4d8d07b63de5 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@nxp.com>
Date: Tue, 22 May 2018 15:31:07 +0800
Subject: [PATCH 3867/5242] MLK-18368-3: hdp: add
 CDN_API_InfoframeRemovePacket function

commit  1eb275903ad4fd07eaa8a3250e7da0be0c5fe75f from
https://source.codeaurora.org/external/imx/linux-imx.git

add function to remove infoframe base on the packet type.

Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/hdp/imx-hdp-audio.c |    4 ++++
 drivers/mxc/hdp/API_Infoframe.c         |   17 +++++++++++++++++
 drivers/mxc/hdp/API_Infoframe.h         |    1 +
 3 files changed, 22 insertions(+)

diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp-audio.c b/drivers/gpu/drm/imx/hdp/imx-hdp-audio.c
index 05cfe89..48bcc1e 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp-audio.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp-audio.c
@@ -165,6 +165,10 @@ static int imx_hdp_audio_hw_params(struct device *dev, void *data,
 
 static void imx_hdp_audio_shutdown(struct device *dev, void *data)
 {
+	struct imx_hdp *hdmi = dev_get_drvdata(dev);
+	state_struct *state = &hdmi->state;
+
+	CDN_API_InfoframeRemovePacket(state, 0x1, 0x84);
 }
 
 static int imx_hdp_audio_get_eld(struct device *dev, void *data, uint8_t *buf, size_t len)
diff --git a/drivers/mxc/hdp/API_Infoframe.c b/drivers/mxc/hdp/API_Infoframe.c
index eb6c5e8..2a4e973 100644
--- a/drivers/mxc/hdp/API_Infoframe.c
+++ b/drivers/mxc/hdp/API_Infoframe.c
@@ -144,3 +144,20 @@ CDN_API_STATUS CDN_API_InfoframeRemove(state_struct *state, u8 entry_id)
 
 	return CDN_OK;
 }
+
+CDN_API_STATUS CDN_API_InfoframeRemovePacket(state_struct *state, u8 entry_id, u8 packet_type)
+{
+	/* invalidate entry */
+	if (cdn_apb_write
+	    (state,
+	     BANK_OFFSET | ADDR_SOURCE_PIF | (SOURCE_PIF_PKT_ALLOC_REG << 2),
+	     0x20000 | F_PKT_ALLOC_ADDRESS(entry_id) |  F_PACKET_TYPE(packet_type)))
+		return CDN_ERR;
+	if (cdn_apb_write
+	    (state,
+	     BANK_OFFSET | ADDR_SOURCE_PIF | (SOURCE_PIF_PKT_ALLOC_WR_EN << 2),
+	     F_PKT_ALLOC_WR_EN(1)))
+		return CDN_ERR;
+
+	return CDN_OK;
+}
diff --git a/drivers/mxc/hdp/API_Infoframe.h b/drivers/mxc/hdp/API_Infoframe.h
index d8811e7..492eb66 100644
--- a/drivers/mxc/hdp/API_Infoframe.h
+++ b/drivers/mxc/hdp/API_Infoframe.h
@@ -59,5 +59,6 @@ CDN_API_STATUS CDN_API_InfoframeSetNoActiveIdle(state_struct *state,
 						u8 entry_id, u8 packet_len,
 						u32 *packet, u8 packet_type);
 CDN_API_STATUS CDN_API_InfoframeRemove(state_struct *state, u8 entry_id);
+CDN_API_STATUS CDN_API_InfoframeRemovePacket(state_struct *state, u8 entry_id, u8 packet_type);
 
 #endif
-- 
1.7.9.5

